File: 0032. Longest Valid Parentheses/Solution0032.cpp
class Solution0032 {
public:
    int longestValidParentheses(string s) {
        int left = 0, right = 0, maxLen = 0, n = s.length();
        for (int i = 0; i < n; i++) {
            if (s[i] == '(') {
                left++;
            } else if (s[i] == ')') {
                right++;
            }
            if (left == right) {
                maxLen = max(maxLen, right * 2);
            } else if (right > left) {
                left = 0; right = 0;
            }
        }
        left = 0; right = 0;
        for (int i = n-1; i >= 0; i--) { // ()(()
            if (s[i] == '(') {
                left++;
            } else if (s[i] == ')') {
                right++;
            }
            if (left == right) {
                maxLen = max(maxLen, left * 2);
            } else if (right < left) {
                left = 0; right = 0;
            }
        }
        return maxLen;
    }
};

class Solution0032_1 {
public:
    int longestValidParentheses(string s) {
        int maxLen = 0;
        stack<int> stk;
        int n = s.length();
        for (int i = 0; i < n; i++) {
            if (s[i] == '(') {
                stk.push(i);
            } else {
                if (!stk.empty() && s[stk.top()] == '(') {
                    stk.pop();                
                    int left = stk.empty()? -1 : stk.top();
                    maxLen = max(maxLen, i - left);
                } else {
                    stk.push(i);
                }
            }
        }
        return maxLen;
    }
};

File: 0032. Longest Valid Parentheses/Solution0032.java
public class Solution0032 {
    public int longestValidParentheses(String s) {
        int left = 0, right = 0, maxlength = 0;
        for (int i = 0; i < s.length(); i++) {
            if (s.charAt(i) == '(') {
                left++;
            } else {
                right++;
            }
            if (left == right) {
                maxlength = Math.max(maxlength, 2 * right);
            } else if (right > left) {
                left = right = 0;
            }
        }
        left = right = 0;
        for (int i = s.length() - 1; i >= 0; i--) {
            if (s.charAt(i) == '(') {
                left++;
            } else {
                right++;
            }
            if (left == right) {
                maxlength = Math.max(maxlength, 2 * left);
            } else if (left > right) {
                left = right = 0;
            }
        }
        return maxlength;
    }
}

// class Solution { // 05 - 08
//     // public int longestValidParentheses(String s) {
//     //     int max = 0;
//     //     Deque<Integer> stack = new ArrayDeque<>();
//     //     for (int i = 0; i < s.length(); i++) {
//     //         char c = s.charAt(i);
//     //         if (c == '(') {
//     //             stack.offerLast(i);
//     //         } else {
//     //             if (!stack.isEmpty() && s.charAt(stack.peekLast()) == '(') {
//     //                 stack.pollLast();
//     //                 int left = stack.isEmpty()? -1 : stack.peekLast();
//     //                 max = Math.max(max, i - left); // (left, right]
//     //             } else {
//     //                 stack.offerLast(i);
//     //             }
//     //         }
//     //     }
//     //     return max;
//     // }

// }

/** 
max:2-0=2,3-0=3
stk:

   i
0123
(())

*/
